### ISSUE TITLE
Incremental crypto_lock/crypto_sign API

### ISSUE THREAD

**Issue by CuleX** [LoupVaillant/Monocypher#29]
It may be worthwhile to also create an incremental `crypto_lock`/`crypto_unlock` API for cases where you have multiple buffers to decrypt (e.g. header and body in separate buffers) or a large file to read out in chunks.

Same for `crypto_sign`/`crypto_check` (think: signing fairly large update files).

YMMV, but it's an idea.

**Comment by CuleX** [comment]
# crypto_lock

> `crypto_unlock` highlights a dilemma: a one pass incremental interface has to interleave authentication and decryption. If the message is corrupted, you lost 3 times more time than you would have if you only did the authentication.

Isn't that a tradeoff you can shift to the user instead? I'm considering the following situations:

* Multiple medium-sized buffers in memory where allocating one for the whole message may risk violating memory constraints. If you're dealing with corrupted memory, the lost time is the least of your issues.
* One very large file read in chunks (where `mmap()` is not an option e.g. because you're on a 32-bit platform without PAE and the file's beyond what you could map in a 32-bit address space, or because you're looking at an embedded platform whose kernel *can't* `mmap()`), primarily from at-rest AE. The only failure methods I can think of are an attacker (who's clearly got no sane plan) and random corruption. The latter is, again, quite a lot more grave than the other, dwarving the performance of your call. The former is, depending on your threat model, unlikely enough that you can sacrifice the performance for the memory cost improvements.
* Multiple individual small messages on the network. *This* is where your performance loss becomes frustrating. I believe TCP can genreally be assumed to be reliable these days and corruption is fairly likely an attempt at an attack. There's a tradeoff depending on the threat model since you're probably going to shortcut a failure here, and that's IMHO best decided by the user, rather than the library.

---

# crypto_sign

## Solution involving documentation

> The situation is somewhat reversed for signing. Signature verification only requires one pass, so an incremental interface is easy. To sign a message however, you have to hash it twice. 

Okay, who thought this was a good idea?

Neither of the two options to sidestep this are very useful. I'd settle for a note in the documentation, such as: "The algorithm used requires the full message to be hashed twice at separate times. Workarounds would involve lowering the security guarantees of the underlying hash function or introduce high potential for misuse, likely leading to situations such as the Sony PlayStation EdDSA nonce randomization failure. For this reason, an incremental interface has not been provided; prefer signing a hash of your data instead, see `crypto_blake2b_init(3monocypher)` for details about incremental hashing." under the Implementation Details section.

## Solution in code possible after all?

However, may I propose a different solution? **Disclaimer**: I am not familiar with EdDSA or elliptic curves in general.

Looking at `crypto_sign`, the two hash passes seem to be generating `R` from HASH(HASH(secret_key)[31..63] || message) and `h_ram` from HASH(message). What if `crypto_sign_ctx` looked something like this:

```c
typedef struct {
    crypto_blake2b_ctx hash_R;
    crypto_blake2b_ctx hash_msg;
} crypto_sign_ctx;```

with `crypto_sign_init(crypto_sign_ctx *ctx, const uint8_t secret_key[32], const uint8_t public_key[32])` initializing the prefix in `hash_R`. `crypto_sign_update(crypto_sign_ctx *ctx, const uint8_t *message, size_t message_size)` then feeds the message to both hashes in `ctx`. `crypto_sign_final(crypto_sign_ctx *ctx, uint8_t signature[64])` doing the actual EdDSA using the hashes from `ctx`. I fail to see why this wouldn't work, but there may be subtleties I'm not seeing.

**Comment by LoupVaillant** [comment]
Crypto_lock
---------------

> Isn't that a tradeoff you can shift to the user instead? 

Yes it is. We just provide the incremental interface, and the user can then use it, possibly like this:

    crypto_lock_ctx;
    crypto_lock_init(&ctx);
    crypto_lock_encrypt(&ctx, msg1, msg_size1);
    crypto_lock_encrypt(&ctx, msg2, msg_size2);
    crypto_lock_encrypt(&ctx, msg3, msg_size3);
    crypto_lock_authenticate(&ctx, tag);

    crypto_unlock_ctx;
    crypto_unlock_init(&ctx);
    crypto_unlock_decrypt(&ctx, msg1, msg_size1);
    crypto_unlock_decrypt(&ctx, msg2, msg_size2);
    crypto_unlock_decrypt(&ctx, msg3, msg_size3);
    if (crypto_unlock_verify(&ctx, tag)) {
        corrupted_message();
    } else {
        all_is_well();
    }

(We could use "final" instead of "authenticate" and "verify" here for consistency, but I feel this would obscure the purpose of those functions —which by the way could be skipped if the user is moronic enough.)

crypto_sign
---

> Okay, who thought this was a good idea?

Daniel J. Bernstein Niels Duif, Tanja Lange, Peter Schwabe, and Bo-Yin Yang. They explain their reason in the [Ed25519 paper](http://ed25519.cr.yp.to/ed25519-20110926.pdf) (search for _Pseudorandom Generation of r_). They even mention the Sony PS3 disaster.

> Neither of the two options to sidestep this are very useful. I'd settle for a note in the documentation, such as: […]

Err, your note precisely describes the first workaround (first hash the message, then sign the hash). Otherwise it's perfect. Including it in the man page is a good idea, a pull request would be welcome.

> However, may I propose a different solution?

Alas, the data flow prevents that:

    R = Base_point × Hash(private_key | message);
    H = Hash(R | public_key | message);

To compute the second hash, you need the result of the first hash, hence the 2 passes. Now if EdDSA had been defined this way instead:

    R = Base_point × Hash(private_key | message);
    H = Hash(public_key | message | R);

Then you could have performed the hashes in parallel. If it had been designed *that* way:

    R = Base_point × Hash(message | private_key);
    H = Hash(message | R | public_key);

Then you'd only have to hash the message once (just copy the context to finalise the hash differently in both cases. Problem is, I'm not sure how this would affect collision resilience, or worse, length collision attacks. If the EdDSA inventors thought of the issue, their settling on the less efficient solution is pretty strong evidence my proposals are bogus in some way.  I'm not going to invent that particular bit of crypto without serious expert analysis.

**Comment by CuleX** [comment]
> (We could use "final" instead of "authenticate" and "verify" for consistency, but I feel this would obscure the purpose of those functions —which by the way could be skipped if the user is moronic enough.)

At least that's how libgcrypt does it, having a `gcry_cipher_checktag` function. The naming is becoming kind of a problem borderlining bikeshed: "verify" might suggest it's just kind of "optional", whereas "final" sounds more like it's "required". I'm fine with either naming, but fairly adamant about having the interface.

---

> Daniel J. Bernstein Niels Duif, Tanja Lange, Peter Schwabe, and Bo-Yin Yang. They explain their reason in the Ed25519 paper (search for Pseudorandom Generation of r). They even mention the Sony PS3 disaster.

Oh for heaven's sake. I'm sure there's a reason why they went with this, though. It's kind of hard not to notice the effects of designing it like this.

> Err, your note precisely describes the first workaround (first hash the message, then sign the hash). Otherwise it's perfect. Including it in the man page is a good idea, a pull request would be welcome.

There are a number of other things I've noticed/wanted in the documentation while writing some code that uses Monocypher, so I'll just go ahead and add that in there.

Thanks for clearing up my misconception. I skimmed the code too quickly, failing to realize that R gets used in the same place the message is hashed; I apologize.

**Comment by LoupVaillant** [comment]
> I skimmed the code too quickly, failing to realize that R gets used in the same place the message is hashed; I apologize.

No problem :smile:, looking at the code, I can see why you missed it. Placing R in `signature` before hashing it wasn't the most readable thing to do. I'll think of a more readable way if possible, or perhaps add a small comment.

**Comment by CuleX** [comment]
Requesting  confirmation that the resolution for **lock and unlock is to provide an incremental interface and document the tradeoff.**

---

> Yeah, DJB in particular is very performance sensitive. Still, this might be worth investigating, just in case.

Out of curiosity, how do you intend to investigate? Sending an e-mail to the paper's authors and pray it gets a response?

**Comment by LoupVaillant** [comment]
> Requesting confirmation that the resolution for lock and unlock is to provide an incremental interface and document the tradeoff.

Confirmed. I will write an incremental interface for lock and unlock before the next official release. We will then close this issue.

I'm still not sure about `crypto_check()`, so let's put it on hold for now. I may add it later if I'm presented with a credible use case.

---

> Out of curiosity, how do you intend to investigate?

First, ask on Reddit/r/crypto. Maybe on other forums too. Then, if it looks like the authors may have overlooked something, try and contact the paper's authors. It's a pretty long shot, but you never know. I *did* manage to report an Argon2i bug just through r/crypto.

Another route would be to write a nice blog post that would include this bummer, have it gather karma and comments on HN. I think I'll have enough material to write a good click bait by the time we publish the next release.

**Comment by LoupVaillant** [comment]
Hi, I have just sketched out an incremental interface, with the following API. What do you think?

    void crypto_aead_lock_init(crypto_lock_ctx *ctx,
                               const uint8_t key  [32],
                               const uint8_t nonce[24],
                               const uint8_t *ad, size_t ad_size);
    
    void crypto_lock_init(crypto_lock_ctx *ctx,
                          const uint8_t key[32],
                          const uint8_t nonce[24]);
    
    void crypto_lock_update(crypto_lock_ctx *ctx,
                            uint8_t *cipher_text,
                            const uint8_t *plain_text, size_t text_size);
    
    void crypto_lock_final(crypto_lock_ctx *ctx, uint8_t mac[16]);
    
    void crypto_unlock_update(crypto_lock_ctx *ctx,
                              uint8_t *plain_text,
                              const uint8_t *cipher_text, size_t text_size);
    
    int crypto_unlock_final(crypto_lock_ctx *ctx, const uint8_t mac[16]);

(Note: there is no `crypto_aead_unlock_init` nor `crypto_unlock_init` because they would be the same as their `lock` counterparts —a bit like how Chacha20 decryption is the same as encryption.)

**Comment by CuleX** [comment]
Thoughts on this bundle of changes?

1. Remove `crypto_aead_lock_init` in favor of `crypto_lock_init`.
2. Add `crypto_aead_ad_update`. It would need to be documented that you have to process `ad` before any data to be encrypted.

**Rationale**: The same as for an incremental ciphertext API; your data could be in various different buffers. **Downside**: Reading the documentation is required for correct usage since processing `ad` after your plaintext is equally intuitive to processing it before your plaintext.

I have no strong feelings about this, so if you feel downside <= rationale, keep as-is.

---

Everything else looks just about how I'd expect it to. Make sure to keep the guarantees for overlapping `plain_text` and `cipher_text`.

Unfortunately, there's no good way for the `final` functions to be agnostic to the underlying cipher being a stream cipher. You cannot change Monocypher to use a block cipher with this API without breaking API compatibility because the `final` functions cannot apply padding. That said, I do not expect Monocypher to switch to a block cipher at any point in the future.

**Comment by LoupVaillant** [comment]
Thanks for the feedback. Your bundle of changes is enticing. I had originally decided against incremental aead processing because I figured there'd be *no fucking way* someone could come up with a legitimate use case for huge amounts of additional data —as I often say, if it's worth authenticating, it's probably worth encrypting.

Still, your change introduce a nice orthogonality. I think I'll take it. The downside is not so problematic: if you want to authenticate your additional data after encryption, you can. Heck, you could even *interleave* calls to `crypto_lock_update` and `crypto_lock_ad_update` if you felt like it. It wouldn't be compatible with the monolithic interface, but the security guarantees will hold. Incorrect usage is possible, but not dangerous.

---

Beware of overlapping buffers. I believe (but need to test properly) that key and nonce buffers can overlap with anything else. The plaintext and cyphertext however must be either equal, or *disjoint*. Partial overlaps may give you wrong results.

---

I don't think Monocypher will ever use a block cipher. Even if it does, there's still CTR mode, which builds a stream cipher on top of a block cipher —no padding needed. In any case, Monocypher strongly depends on its underlying primitives. Any change there will trigger a general overhaul, and an incompatible, new version of Monocypher.

**Comment by LoupVaillant** [comment]
Done at last. I quickly cobbled together a piece of documentation for the incremental interface, you may want to have a more serious look at it.
